src/wayland/data_device/mod.rs

```diff
@@ -1,313 +1,117 @@
+use super::seat::Seat;
 
-static DND_ICON_ROLE: &str = "dnd_icon";
+mod device;
+mod dnd_grab;
+mod seat_data;
+mod source;
 
-/// Events that are generated by interactions of the clients with the data device
-#[derive(Debug)]
-pub enum DataDeviceEvent {
-    /// A client has set the selection
-    NewSelection(Option<wl_data_source::WlDataSource>),
-    /// A client started a drag'n'drop as response to a user pointer action
-    DnDStarted {
-        /// The data source provided by the client
-        ///
-        /// If it is `None`, this means the DnD is restricted to surfaces of the
-        /// same client and the client will manage data transfer by itself.
-        source: Option<wl_data_source::WlDataSource>,
-        /// The icon the client requested to be used to be associated with the cursor icon
-        /// during the drag'n'drop.
-        icon: Option<wl_surface::WlSurface>,
-        /// The seat on which the DnD operation was started
-        seat: Seat,
-    },
-    /// The drag'n'drop action was finished by the user releasing the buttons
-    ///
-    /// At this point, any pointer icon should be removed.
-    ///
-    /// Note that this event will only be generated for client-initiated drag'n'drop session.
-    DnDDropped {
-        /// The seat on which the DnD action was finished.
-        seat: Seat,
-    },
-    /// A client requested to read the server-set selection
-    SendSelection {
-        /// the requested mime type
-        mime_type: String,
-        /// the fd to write into
-        fd: RawFd,
-    },
+pub trait DataDeviceHandler: Sized {
+    fn data_device_state(&self) -> &DataDeviceState<Self>;
 }
 
-enum Selection {
-    Empty,
-    Client(wl_data_source::WlDataSource),
-    Compositor(SourceMetadata),
+pub struct DataDeviceState<D> {
+    log: slog::Logger,
+    manager_global_id: GlobalId,
+    callback: Arc<Mutex<dyn FnMut(DataDeviceEvent<D>) + Send + 'static>>,
+    action_choice: Arc<Mutex<dyn FnMut(DndAction, DndAction) -> DndAction + Send + 'static>>,
 }
 
-struct SeatData {
-    known_devices: Vec<wl_data_device::WlDataDevice>,
-    selection: Selection,
-    log: ::slog::Logger,
-    current_focus: Option<Client>,
+impl<D> std::fmt::Debug for DataDeviceState<D> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DataDeviceState")
+            .field("manager_global_id", &self.manager_global_id)
+            .finish_non_exhaustive()
+    }
 }
 
-impl SeatData {
-    fn set_selection(&mut self, new_selection: Selection) {
-        self.selection = new_selection;
-        self.send_selection();
-    }
+impl<D> DataDeviceState<D> {
+    pub fn new<F, C, L>(display: &mut Display<D>, callback: C, action_choice: F, logger: L) -> Self
+    where
+        L: Into<Option<::slog::Logger>>,
+        D: GlobalDispatch<WlDataDeviceManager, GlobalData = ()> + 'static,
+        D: DataDeviceHandler,
+        F: FnMut(DndAction, DndAction) -> DndAction + Send + 'static,
```

Comment:
These two callbacks could be defined in a trait function?

Reply:
I initially tried to do so but failed, they are called from grabs.

Reply:
oh, wait nevermind, we have access to D in grabs now. I guess I will try again.

Reply:
Yep, enforcing D in grabs allows for those to be moved to trait. :rocket: :rocket: :rocket: :rocket: :rocket: